test/unit/test_il_vm.c

```diff
@@ -0,0 +1,147 @@
+// SPDX-FileCopyrightText: 2021 heersin <teablearcher@gmail.com>
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#include <rz_il.h>
+#include <rz_util.h>
+#include "minunit.h"
+
+static bool test_rzil_vm_init() {
+	RzILVM *vm = RZ_NEW0(struct rz_il_vm_t);
+	mu_assert_notnull(vm, "Create VM");
+	rz_il_vm_init(vm, 0, 8, 8);
+	mu_assert_eq(vm->addr_size, 8, "VM Init");
+	rz_il_vm_close(vm);
+	mu_end;
+}
+
+static bool test_rzil_vm_basic_operation() {
+	RzILVM *vm = RZ_NEW0(struct rz_il_vm_t);
+	mu_assert_notnull(vm, "Create vm");
+	rz_il_vm_init(vm, 0, 8, 16);
+
+	// 1. create variables
+	RzILVar *var_r1 = rz_il_vm_create_variable(vm, "r1");
```

Comment:
How can you assign a type to a `RzILVar`? Also, why is the type only in the `RzILVal` and not in the `RzILVar`? Is it possible this way to assign different `RzILVal` types to the same `RzILVar`?

Reply:
There is a a clear distinction between variables and values, see:
- http://binaryanalysisplatform.github.io/bap/api/odoc/bap-core-theory/Bap_core_theory/Theory/Value/index.html
- http://binaryanalysisplatform.github.io/bap/api/odoc/bap-core-theory/Bap_core_theory/Theory/Var/index.html

On the other hand, you are right, the sorts of Vars and Variables are related:
```ocaml
val sort : 'a t -> 'a Value.sort
(*  sort v is variable's sort. *)
```
You can find all sorts of the Value here: http://binaryanalysisplatform.github.io/bap/api/odoc/bap-core-theory/Bap_core_theory/Theory/Value/Sort/index.html

> Module Value.Sort

Value Sorts.

A concrete and extensible representation of a value sort. The sort usually holds the static information about the value representation, like the width of a bitvector, the representation format of a floating point number, and so on.

This module is mostly needed when a new sort is defined. The Core Theory provides a predefined collection of sorts, here is the list:

    Bitv - bitvectors, e.g., BitVec(i)
    Mem - memories, e.g., Mem(BitVec(i), BitVec(j)
    Float - floating points, e.g., Float(IEEE754(2, 8, 23), BitVec(32);
    Rmode - rounding mode, e.g., Rmode.

This module defines a simple DSL for specifying sorts, the DSL grammar is made only from three rules:

           sort = sym | int | sort(sort)

The DSL is embedded into the host language with an infix operator @-> for application, e.g., OCaml grammar for sorts is:

v sort = sym exp | int exp | sort "@->" sort exp = ?a valid OCaml expression? v

Both symbols and numbers are indexed with a type index, which serves as a witness of the sort value, e.g.,

type int8
let int8 : int8 num sort = Sort.int 8

Type indices enable explicit reflection of the target language type system in the host type system, while still keeping the typing rules under designer's control.